### DISCUSSION TITLE
Backport `Sidekiq::JobUtil#verify_json` error message to v6.x

### DISCUSSION THREAD

**Discussion by jai-x** [sidekiq/sidekiq#6263] [Ideas]
In this PR (https://github.com/sidekiq/sidekiq/pull/5786), a very useful change was made to the error message when a non-JSON serializable argument was passed to job. The error message includes the `unsafe_item.inspect` and `unsafe_item.class` in the error message, however this is only available in v7.

I'm in the middle of trying diagnose a mysterious occurance of this error message in a v6 install of Sidekiq.
I have raising enabled for non-JSON args, but finding it hard to pinpoint what is causing this in my app.

It would be great if this error message, including the `inspect` and `class` outputs for the unsafe item were backported to v6.
It would make upgrading to v7 a lot easier in my case.

**Comment by mperham** [comment]
I consider Sidekiq 6 frozen except for high-priority compatibility fixes or security issues. Can you monkeypatch the method in your initializer?

**Reply by jai-x** [reply]
Sure, I've ended up with a monkeypatch solution.

For reference it's this:

```ruby
# in file: config/initializers/sidekiq.rb

Sidekiq.strict_args!

Sidekiq.configure_server do |config|
   # ...
end

class Sidekiq::Client
  def verify_json(item)
    job_class = item["wrapped"] || item["class"]
    args = item["args"]
    mode = Sidekiq[:on_complex_arguments]

    if mode == :raise || mode == :warn
      if (unsafe_item = json_unsafe?(args))
        msg = <<~EOM
          Job arguments to #{job_class} must be native JSON types.
          Argument #{unsafe_item.inspect} is a #{unsafe_item.class}.
          All arguments to job: #{args.inspect}
          See https://github.com/sidekiq/sidekiq/wiki/Best-Practices
          To disable this error, add `Sidekiq.strict_args!(false)` to your initializer.
        EOM

        if mode == :raise
          raise(ArgumentError, msg)
        else
          Sidekiq.logger.warn(msg)
        end
      end
    end
  end

  private

  RECURSIVE_JSON_UNSAFE = {
    Integer => ->(val) {},
    Float => ->(val) {},
    TrueClass => ->(val) {},
    FalseClass => ->(val) {},
    NilClass => ->(val) {},
    String => ->(val) {},
    Array => ->(val) {
      val.each do |e|
        unsafe_item = RECURSIVE_JSON_UNSAFE[e.class].call(e)
        return unsafe_item unless unsafe_item.nil?
      end
      nil
    },
    Hash => ->(val) {
      val.each do |k, v|
        return k unless String === k

        unsafe_item = RECURSIVE_JSON_UNSAFE[v.class].call(v)
        return unsafe_item unless unsafe_item.nil?
      end
      nil
    }
  }

  RECURSIVE_JSON_UNSAFE.default = ->(val) { val }
  RECURSIVE_JSON_UNSAFE.compare_by_identity
  private_constant :RECURSIVE_JSON_UNSAFE

  def json_unsafe?(item)
    RECURSIVE_JSON_UNSAFE[item.class].call(item)
  end
end

```

**Reply by mperham** [reply]
Exactly. I would move it to its own file and add a guard up top which raises an error if Sidekiq::VERSION > "7" since you don't want to patch v7 when you upgrade.
